contracts/os/andromeda-kernel/src/execute.rs

```diff
@@ -797,23 +798,68 @@ impl MsgHandler {
                 error: Some(format!("Channel not found for chain {chain}")),
             });
         }?;
-        let ctx = ctx.map_or(
-            AMPPkt::new(
-                info.sender,
-                env.clone().contract.address,
-                vec![AMPMsg::new(
-                    recipient.clone().get_raw_path(),
-                    message.clone(),
-                    None,
-                )],
-            ),
+        let ctx = ctx.map_or_else(
+            || {
+                // Check if sender has username
+                let vfs_address = KERNEL_ADDRESSES.load(deps.storage, VFS_KEY).unwrap();
+                let msg = VFSQueryMsg::GetUsername {
+                    address: info.sender.clone(),
+                };
+                let username_addr =
+                    deps.querier
+                        .query::<String>(&cosmwasm_std::QueryRequest::Wasm(
+                            cosmwasm_std::WasmQuery::Smart {
+                                contract_addr: vfs_address.to_string(),
+                                msg: to_json_binary(&msg).unwrap(),
+                            },
+                        ));
+                match username_addr {
+                    Ok(username) => {
+                        if username != info.sender {
+                            AMPPkt {
+                                messages: vec![AMPMsg::new(
+                                    recipient.clone().get_raw_path(),
+                                    message.clone(),
+                                    None,
+                                )],
+                                ctx: AMPCtx::new(
+                                    info.sender,
+                                    env.clone().contract.address,
+                                    0,
+                                    Some(AndrAddr::from_string(username)),
+                                ),
```

Comment:
The only thing that should be changing in the packet is the context's username reference, may be a cleaner way to achieve this.